Solve or create note with photo

ENT8R opened the pull request:
This PR is the updated version of #455 which fixes #90.
This time the camera app opens via intent, the user takes a photo which is stored in the local storage of the phone. Then, if the note should be uploaded, the photos are uploaded and **then** a text with the link to the images will be attached to the note text or comment.
I decided against uploading the images to  https://lut.im because the images are only kept for 365 days, even if I want to store it longer, so the images are uploaded to https://images.mondedie.fr which is another instance of lut.im. Maybe it would be better to setup our **own** instance of lut.im to have the full control of how long the images are kept on the server and so on...
I had a few diffuculties with implementing the feature for the note discussion quest: I don't know why, but the array which points to the images is always NULL in the database, even if I attached a photo... For a new created note it works as expected (The array with the paths to the images on local storage is stored in the database and is not always NULL). Maybe you can find the issue...

### Here are some screenshots:
### Creating a new note:
<img src="https://user-images.githubusercontent.com/25306497/31789063-1c33001a-b511-11e7-9ef3-9804c4b86cab.png" width="300px">    <img src="https://user-images.githubusercontent.com/25306497/31789067-1f0c9080-b511-11e7-92e4-dfc196e0a37c.png" width="300px">

### Taking part in a note discussion:
<img src="https://user-images.githubusercontent.com/25306497/31789069-228008fa-b511-11e7-80e6-2707e2ea224f.png" width="300px">    <img src="https://user-images.githubusercontent.com/25306497/31789071-24ba8f78-b511-11e7-8a1b-7274eb77c446.png" width="300px">

westnordost commented:
Hui, wow, that's a lot. Thank you for starters for being persistent about this feature :-)

I congratulate you for tackling such a complex feature - with saving to DB, saving to local storage, getting picture from camera and finally uploading the pic to some service.

I will have to look through this more thoroughly (on the weekend), but scanning throw it quickly, I found the following issues:

1. UI Issue: [Attach Photo] should not be a button in the button panel next to [OK] and [No, Hide] because the expectation for these buttons is that they are a definite action (i.e. that they close the panel). Attaching a photo does (should not) close the panel - you can still write something thereafter (I hope). So, the [Attach Photo] button should simply be a normal button below the input field and above panel. If you want to be extra creative, you could align the button in the same way the photos are, when they are displayed after taking them.
2. Write external storage permission: StreetComplete already writes files to storage, specifically the quests.png it generates since recently (see TangramQuestSpriteSheetCreator) and saves it to the private local storage. The created photographs should be saved to there as well to not need that permission 
3. Permission issue: If the app is only asking [the camera app] to take a photo, why does it itself need the camera permission then?
4. Possible save problem: You have an array list and call toString() on it, expect it to be JSON. I'd say, it is just coincidence that this outputs valid JSON, I wouldn't count on it.  (Also, see point 1 of notes)
5. Poor man's lock: In CreateNoteUpload.java line 158, you use a quite creative way to wait for another task to complete before continuing in that thread. It ties up the CPU at 100%, that's not desirable. Basically, there is no need for the ImageUploadHelper to even work asynchronously. **CreateNoteUpload is already executed in a background thread, so ImageUploadHelper could do everything synchronously (no AsyncTask)**. This also simplifies the code. Anyway, **just for your info**, there are several proper ways on how to wait for another asynchronous job to finish, you can read them each up if you like (too complex to explain them in each one bullet point:
   * If you have a reference to the thread for which the "main" thread should wait, you can use `thatOtherThread.join();`.
   * In a general situation, (i.e. yours,) you can use a simple monitor object (`Object monitor= new Object();` .... `monitor.wait()` in "main" thread, `monitor.notify()` in other thread, but watch out for race conditions)
   * Then, there is `CyclicBarrier `and `CountdownLatch`, but these constructs from the java concurrent package are made for more complex use cases.
6. Failing tests: You changed the signature of the NoteDao stuff. That should lead to some failing tests. Did you run the tests and the androidTests? Also, perhaps it would be easily possible to add some tests for the new functionality? (Not sure)

And some notes:
1. I usually use not JSON but Kryo to serialize/deserialize blobs like arrays of strings to/from the database. I would prefer to not mix these techniques.
2. Needless OkHttp3 dependency: Wouldn't HttpConnection do the job as well? The main advantage of OkHttp as I remember is that it by default reuses the socket and does some caching, so it is faster for repeated/frequent requests. But uploading that picture is just one request. (And btw, as far as I know, GZIPing the data is also default for the normal HttpConnection, at least on Android)
3. The URL of the lutim-server should probably not be hardcoded in that class, instead be passed to ImageUploadHelper (actually, I would rename it to LutimImageUploader) and finally be injected to CreateNoteUpload. (Reason: So that the actual URL can be exchanged for unit tests)
4. The privacy note should mention that images are uploaded to that server. Since it is always problematic to change a string (because translations), perhaps add another `privacy_html_addendum1` string that shortly mentions that.

westnordost commented:
Also, regarding that lut.im instance, it has only seen 2000-something uploads since 2014 it seems. Not sure if it would endure such an added burden.
I think, in any case, whichever lutim instance is used in the end, we should first contact the operator and ask for permission to use it for this app I think. Just out of respect and to hear if there are concerns whether the server will be able to take the additional traffic.

Phyks commented:
@ENT8R Concerning the Lutim instance, I know lut.im just had some load issues as it was used by some Kodi apps to serve TV icons (CDN-like) and it could not handle the load. As most of Lutim instances are run by benevolents, it might be worth checking they can handle the load or run StreetCpmplete's own. 

Looks like an amazing feature otherwise!

antonv6 commented:
Since this feature heavily depends on a service that could host these attached photos, I'd call that the main issue. Maybe it's not out of scope of an Android app to set up its own file hosting server, but let's remember that there needs to be someone committed enough to maintain the thing just for this feature.

What about OpenStreetCam and Mapillary?

westnordost commented:
@ENT8R I read through the Android documentation for taking photos and corrected my initial post.

ENT8R commented:
I am trying to fix all the things you mentioned right now. But I implemented 2. (write external storage permission) already and it works as expected...

rugk commented:
Maybe provide a good default and let the user configure the server, if they want to change it (because it is down or they want to use their own). You may also provide some nice big servers as "suggestions". (I imagine a dropdown with suggestions and manual input, similar to like you can enter the street name.)

Or you automate that and have a list of servers and randomly upload to one server out of that. Or at least fallback…

ENT8R commented:
> Permission issue: If the app is only asking [the camera app] to take a photo, why does it itself need the camera permission then?

The app does not request the camera permission it is just a `uses-feature` tag, but I think this is not necessary, so I will remove this with the next commit.

ENT8R commented:
With the latest commit issue 1,2,3 and 5 is fixed and all of your notes about the code are fixed too.
I still can't find the issue behind 4... maybe you can find out more if you look at the code... Regarding point 6: I will try to do this this weekend.

westnordost commented:
Hmm? Did you notice that I revoked what I said for issues 2 and 3 after I read the official Android documentation, realizing that apparently this cannot be done any other way?

westnordost commented:
May I also work on your PR branch? There are some smaller things I noticed where it would be much more efficient if I just changed it rather than telling you to.

ENT8R commented:
> May I also work on your PR branch?

Yes of course! Feel free to fix them!

> Did you notice that I revoked what I said for issues 2 and 3

Yes I noticed it but it works as expected even **without** external storage permission...

westnordost commented:
Also with api version 18?

ENT8R commented:
Mmhh... this is a good point... I can't test it, because I have no devices running on API 18 so I will revert this.

ENT8R commented:
With the latest commit I reverted the way how the file is created and fixed the tests. I did not create new tests (I just added `null` as a parameter to the functions, because I am not familiar at all with tests...).
And finally I fixed issue 4) 🎉, so the path to the photos is now stored correctly in the database if the user takes part in a note discussion and wants to attach a photo.

westnordost commented:
I also changed some stuff and am done for now. I leave this to you again.

ENT8R commented:
I think I'm done now too...
We just have to decide whether we should upload the photos to https://lut.im/ where the images are deleted automatically after one year, but the server seems to be stable or to https://images.mondedie.fr/ where we can store the photos as long as we want but we don't know how stable the server is, because it is not used that much... Or we try out https://framapic.org/ which is also an instance of lut.im with about 400 000 uploaded photos.
I really don't know which server is the best for our purposes...
What do you think?

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/data/osmnotes/CreateNoteDao.java:
```diff
@@ -38,6 +41,10 @@ public boolean add(CreateNote note)
 		}
+		if (note.imagePaths != null)
+		{
+			values.put(CreateNoteTable.Columns.IMAGE_PATHS, note.imagePaths.toString());
```
**Serialization issue:** You need to actually serialize the array list here, not toString() it.

ENT8R replied:
Done

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/data/StreetCompleteOpenHelper.java:
```diff
@@ -115,7 +115,8 @@
 				OsmNoteQuestTable.Columns.LAST_UPDATE + 	" int			NOT NULL, " +
-				OsmNoteQuestTable.Columns.NOTE_ID +			" INTEGER		UNIQUE NOT NULL " +
+				OsmNoteQuestTable.Columns.NOTE_ID +			" INTEGER		UNIQUE NOT NULL, " +
+				OsmNoteQuestTable.Columns.IMAGE_PATHS +		" text " +
```
**Serialization issue:** If you serialize/deserialize the image paths from a byte[] blob (using `serializer`), this needs to be of type `blob`, not `text`.

ENT8R replied:
Done

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/data/StreetCompleteOpenHelper.java:
```diff
@@ -140,7 +141,8 @@
 					CreateNoteTable.Columns.TEXT + 		" text			NOT NULL, " +
-					CreateNoteTable.Columns.QUEST_TITLE + " text" +
+					CreateNoteTable.Columns.QUEST_TITLE + " text, " +
+					CreateNoteTable.Columns.IMAGE_PATHS + " text" +
```
**Serialization issue:** If you serialize/deserialize the image paths from a byte[] blob (using `serializer`), this needs to be of type `blob`, not `text`.

ENT8R replied:
Done

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/util/LutimImageUploader.java:
```diff
@@ -0,0 +1,55 @@
+		}
+		return imageLinks;
+	}
+}
```
This class could use some test case. It could even be a simple test case (not androidTest). If you don't know how to do this, I will after/before finally merging your PR.

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/quests/note_discussion/NoteDiscussionForm.java:
```diff
@@ -162,11 +229,61 @@ private void onClickOk()
+				directory
+		);
+	}
+
```
**Code duplication issue:** `LeaveNoteDialog` and `NoteDiscussionForm` share the same component - this "take photo" button plus the gridview that displays the photos already taken.

This should be put into a separate class.

Since you start an activity for result there, I propose that you put this functionality into a fragment and then add that fragment in the layout xml for both places.

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/quests/note_discussion/NoteDiscussionForm.java:
```diff
@@ -162,11 +229,61 @@ private void onClickOk()
+			if (resultCode == RESULT_OK)
+			{
+				imageBitmaps.add(BitmapFactory.decodeFile(photoFile.getAbsolutePath()));
+				imagePaths.add(photoFile.toString());
```
Doesn't this mean that the whole photo is displayed in the gridview and not just a thumbnail? This could lead to an Out of Memory problem especially with older phones.

ENT8R replied:
And how can we avoid this?

westnordost replied:
Subsampling (reading every x-th pixel into memory). BitmapFactory can do that, see here
https://developer.android.com/reference/android/graphics/BitmapFactory.Options.html#inSampleSize
Look also here: https://stackoverflow.com/questions/4916159/android-get-thumbnail-of-image-on-sd-card-given-uri-of-original-image

westnordost replied:
(I can't tell you exactly how to do this, I'd have to research there as well and I guess this is 90% of the work.)

westnordost replied:
But in my first answer, I meant to paste this link:
https://developer.android.com/topic/performance/graphics/load-bitmap.html#load-bitmap

westnordost replied:
Reading a bit more into the issue, I guess if this becomes more complex than a few dozen lines of code, perhaps simply using Picasso could also be an option.

ENT8R replied:
I am using now this piece of code and it works as expected... ```ThumbnailUtils.extractThumbnail(BitmapFactory.decodeFile(file.getPath()), width, height);```

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/quests/note_discussion/NoteDiscussionForm.java:
```diff
@@ -162,11 +229,61 @@ private void onClickOk()
+					takePhotoIntent.putExtra(MediaStore.EXTRA_OUTPUT, getUriForFile(getActivity(), "de.westnordost.streetcomplete.fileprovider", photoFile));
+				} else {
+					takePhotoIntent.putExtra(MediaStore.EXTRA_OUTPUT, Uri.fromFile(photoFile));
+				}
```
Can you explain what this FileProvider is for? Even after reading the documentation about it, I don't get why it is used here.

ENT8R replied:
The FileProvider is used for getting a content:// URI for the file.
The documentation says:
> Note: We are using ```getUriForFile(Context, String, File)``` which returns a ```content:// URI```. For more recent apps targeting Android 7.0 (API level 24) and higher, passing a ```file:// URI``` across a package boundary causes a ```FileUriExposedException```. Therefore, we now present a more generic way of storing images using a ```FileProvider```.

westnordost replied:
Hmm, so, providing this url to the camera counts as "across package boundary" then, I guess?
Perhaps you could add a comment there with a link to the documentation, just for future reference.

ENT8R replied:
Yes I think so... I will add a link with the next commit

ENT8R replied:
Done

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/quests/note_discussion/NoteDiscussionForm.java:
```diff
@@ -162,11 +229,61 @@ private void onClickOk()
+	{
+		String timeStamp = new SimpleDateFormat("yyyyMMdd_HHmmss").format(new Date());
+		String imageFileName = "JPEG_" + timeStamp + "_";
+		File directory = getActivity().getExternalFilesDir(Environment.DIRECTORY_PICTURES);
```
Doesn't external files dir always require the read device storage permission? The photo should be saved into the app's private storage. (As far as I remember: `getFilesDir`)

ENT8R replied:
The documentation says:
> However, if you'd like the photos to remain private to your app only, you can instead use the directory provided by ```getExternalFilesDir()```. On Android 4.3 and lower, writing to this directory also requires the ```WRITE_EXTERNAL_STORAGE``` permission. Beginning with Android 4.4, the permission is no longer required because the directory is not accessible by other apps, so you can declare the permission should be requested only on the lower versions of Android by adding the ```maxSdkVersion``` attribute

westnordost replied:
Ok

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/data/osmnotes/OsmNoteQuestDao.java:
```diff
@@ -110,6 +116,11 @@
 
+		if (quest.getImagePaths() != null)
+		{
+			values.put(Columns.IMAGE_PATHS, quest.getImagePaths().toString());
```
Serialization issue. See the other comments.

ENT8R replied:
Done

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/data/StreetCompleteOpenHelper.java:
```diff
@@ -302,6 +304,14 @@ public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion)
+			db.execSQL("ALTER TABLE " + OsmNoteQuestTable.NAME + " ADD COLUMN " +
+					OsmNoteQuestTable.Columns.IMAGE_PATHS + " text ;");
+		}
+
```
Also, it should be `blob`, not `text`

ENT8R replied:
> Did you test this as well?

You mean if everything is stored correctly?

westnordost replied:
No, that I'd expect you to do anyway because this is part of the normal "use this feature from beginning to end once" testing.

I mean testing the SQL database upgrade code.

ENT8R replied:
If I upgrade from DB version 7 to 8, a new column called ```image_paths``` is created. I tested this with the "DB Browser for SQLite". Or do you mean something different?

westnordost replied:
Nope, that's fine, thank you! (By the way, for convenience, there are some shell scripts in /tools/ to open it in that DB Browser with one click)

ENT8R replied:
I know... This is how I got to know the browser 😉 👍

ENT8R commented on app/src/debug/res/xml/file_paths.xml:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<paths>
+    <external-path name="images" path="Android/data/de.westnordost.streetcomplete.debug/files/Pictures" />
+</paths>
```
This is needed by the ```FileProvider```, because the ```FileProvider``` needs the directory where the images should be saved. Otherwise it won't work, see https://developer.android.com/reference/android/support/v4/content/FileProvider.html

ENT8R replied:
And it's in debug because the directory is different there. And it's also included in the main folder without the ```.debug``` suffix

westnordost replied:
ok

ENT8R commented:
> b) When the user discards leaving a note / commenting on a note but already uploaded images, the images should be discarded as well.

How can we achieve this with the ```NoteDiscussionForm.java```?

westnordost commented:
> How can we achieve this with the NoteDiscussionForm.java?

Add in `AbstractQuestAnswerFragment.onClickClose` before running the Runnable code like `onDiscard();` with a default empty implementation and override this method in `NoteDiscussionForm`.

ENT8R commented on app/src/main/java/de/westnordost/streetcomplete/data/QuestController.java:
```diff
@@ -255,11 +257,14 @@ else if(quest.getStatus() == QuestStatus.CLOSED)
 	{
 		OsmNoteQuest q = osmNoteQuestDB.get(questId);
+		ArrayList<String> imagePaths = answer.getStringArrayList(NoteDiscussionForm.IMAGE_PATHS);
+		Log.d("imagePaths", imagePaths.toString());
```
Done

ENT8R commented on app/src/main/java/de/westnordost/streetcomplete/quests/AttachPhotoFragment.java:
```diff
@@ -0,0 +1,160 @@
+				imagePaths.add(photoFile.toString());
+
+				LeaveNoteDialog.imagePaths = imagePaths;
+				NoteDiscussionForm.imagePaths = imagePaths;
```
This is not the way how it should be done... I know... but I could not find any other method to pass the array with the image paths to the note creation form or get it from there... Could you help me?

westnordost replied:
Sure. Basically, there is an inter-fragment communication in the app already, you can have a look at how it is done by example there: The `MapFragment` has `MapControlsFragment` as a child fragment.

In a nutshell, the parent fragment uses `getChildFragmentManager().findFragmentById(id);` to get its child and then call methods on it. If communication from the child back to the parent is necessary, the parent can register itself as a listener to the child (also done in the mentioned fragments). But I think this is not necessary for this case.

ENT8R replied:
I implemented it now the same way how it is done in the ```MapFragment``` but if I call ```getChildFragmentManager().findFragmentById(id);``` the result is ```null``` all the time and I get a ```NullPointerException```

westnordost replied:
sure that you are using the right Id? Is the child fragment manager the right one? (Fragment vs Fragment in support library)

ENT8R replied:
This is now fixed

ENT8R commented on app/src/main/java/de/westnordost/streetcomplete/quests/LeaveNoteDialog.java:
```diff
@@ -89,13 +93,15 @@ private void onClickOk()
 	private void onClickCancel()
 	{
+		AttachPhotoFragment attachPhotoFragment = new AttachPhotoFragment();
+		attachPhotoFragment.deleteImages(imagePaths);
```
The same here...

ENT8R commented on app/src/main/java/de/westnordost/streetcomplete/quests/note_discussion/NoteDiscussionForm.java:
```diff
@@ -162,11 +167,20 @@ private void onClickOk()
+	public void onDiscard()
+	{
+		AttachPhotoFragment attachPhotoFragment = new AttachPhotoFragment();
+		attachPhotoFragment.deleteImages(imagePaths);
```
And here too...

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/quests/AbstractQuestAnswerFragment.java:
```diff
@@ -255,10 +255,12 @@ private Resources getEnglishResources()
 		}
 		else
 		{
+			onDiscard();
```
💥 : onDiscard must be in the onYes callback, otherwise it is called **before** the user acknowledges discarding the message.

westnordost commented:
Regarding Thumbnailutils:

Hm yes, I guess you could use that. The problem with this is that to create that thumbnail, still the whole bitmap is loaded into the memory. But of course, this solves at least the primary issue that there are multiple imageviews each displaying a full photo bitmap and thus holding it in memory.

Talking of which, can on taking the photo the camera be asked to output a photo of max a certain resolution and quality?
I am concerned that uploading photos will take a heavy toll on data usage because oftentimes cams output photos in unnecessarily big size (8MB etc when 0,5MB may be more than enough)

ENT8R commented:
> Talking of which, can on taking the photo the camera be asked to output a photo of max a certain resolution and quality?

I did some research and it is not possible at all. The only parameter which the intent accepts is where to store the photo...

westnordost commented:
So, is the PR done from your point of view? If yes, then I will take over to:
- use ThumbnailUtils (I guess) to scale down the image and then overwrite the original photo taken in order to limit the max photo size to around 500KB, instead of i.e. >5M and up. Perhaps I consider putting this as an option in the settings (later).
- create a test for the LutimImageUploader
- contact the administrator of lut.im / alternatives and ask whether it he has concerns if we use his service  for this app

You are invited to do any of that yourself, but you don't have to. I will take care of that then.

ENT8R commented:
Yes I'm done now! Thank you!

westnordost commented:
> contact the administrator of lut.im / alternatives and ask whether it he has concerns if we use his service for this app

Done, waiting for answer.

westnordost commented:
I think I am done now. The last commit was somewhat of a big lump.
There were a few issues with your AttachPhotoFragment that I fixed, then I got carried away - not all changes are fixes.

westnordost commented:
I should have another look if the exception handling is consistent now.

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/data/osmnotes/AttachPhotoUtils.java:
```diff
@@ -0,0 +1,136 @@
+		}
+	}
+
+	public static Bitmap resize(String imagePath, int maxWidth)
```
should probably rename to createScaledBitmap, analogous to the Bitmap static method.

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/quests/LeaveNoteDialog.java:
```diff
@@ -59,6 +61,20 @@ public void onClick(View v)
+		if(savedInstanceState == null)
+		{
+			getChildFragmentManager().beginTransaction().add(R.id.attachPhotoFragment, new AttachPhotoFragment()).commit();
+		}
```
By the way, @ENT8R, what I told you about nested fragments like I did it for MapFragment + MapControlsFragment turns out to be wrong. The Android documentation explicitly states that it is not possible to have nested fragments in layout xmls. Instead, they have to be added programmatically always.

This is only no problem currently with the MapControlsFragment because it never gets removed and added again. Perhaps I should change this for the MapControlsFragment as well.

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/quests/LeaveNoteDialog.java:
```diff
@@ -59,6 +61,20 @@ public void onClick(View v)
+
+	private AttachPhotoFragment getAttachPhotoFragment()
+	{
+		return (AttachPhotoFragment) getChildFragmentManager().findFragmentById(R.id.attachPhotoFragment);
```
Also, I should probably add nullchecks here. Since adding a child fragment is asynchronous, places where that child fragment is used might be called before the fragment is finally added.

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/util/LutimImageUploader.java:
```diff
@@ -0,0 +1,96 @@
+				catch (JSONException e)
+				{
+					Log.e(TAG, "Lutim seems to have a returned a bogus response", e);
+				}
```
I should think of better error handling here.

westnordost commented:
So I put lut.im as image upload service now even though I did not get an answer from the maintainer yet. But it will be some time till the next stable release, so there is still time.

westnordost commented:
He answered me 

> The main problem is not how many photos will be uploaded per month, but how many
> concurrent views these photos will generate. If it's hundreds of views per photo
> concurrently, I'm sorry, but it's not a good idea. If it's tens of view, it can
> be ok.

So, sounds good. I guess the photos uploaded by the app into the notes will actually not really generate concurrent views at all. At least, much less than any pictures uploaded into any forum.

ENT8R commented:
That sounds really great! I don't think that very much people will view an attached photo too. Maybe only the person who wants to fix the note...
